core/src/Cuda/Kokkos_Cuda_Half.hpp

```diff
@@ -127,16 +130,28 @@ KOKKOS_INLINE_FUNCTION
     std::enable_if_t<std::is_same<T, unsigned long long>::value, T>
         cast_from_half(half_t);
 
-class half_t {
+// use alignas(4) to support 32-bit alignment for reductions
+class alignas(4) half_t {
  public:
   using impl_type = Kokkos::Impl::half_impl_t::type;
 
  private:
   impl_type val;
 
  public:
-  KOKKOS_FUNCTION
-  half_t() : val(0.0F) {}
+  // Default constructor - initializes val to 0
+  // noexcept - adds compile time check to ensure exceptions are not thrown
+  // default  - has the compiler define the implicit default constructor even
+  //            though other constructors are present
+  KOKKOS_DEFAULTED_FUNCTION
+  half_t() noexcept = default;
+
+  // Copy constructors
+  KOKKOS_DEFAULTED_FUNCTION
+  half_t(const half_t&) noexcept = default;
+
+  KOKKOS_INLINE_FUNCTION
+  half_t(const volatile half_t& rhs) : val(const_cast<impl_type&>(rhs.val)) {}
```

Comment:
Although this is the first appearance, this is a general comment:

Should be `const_cast<const impl_type&>`, shouldn't it?

I'm not thrilled that we take it by `const volatile&`, then throw away the `volatile`.

Reply:
Agreed. The reason for casting away `volatile` here is that there is no `__half` constructor which accepts a `volatile` RHS. Similarly, on the device, we don't have `__half` operators overloads which accept `volatile` parameters --  Cuda 10 supports `__half` `volatile` stores but not `volatile` arithmetic operands. On the host, I cast away volatileness on the RHS to supress "warning: implicit dereference will not access object of type ‘volatile __half’ in statement".